### PR TITLE
remove all 'message' event listeners when terminating a worker

### PR DESCRIPTION
**Pull request by hhprogram**
solves #147 

 Related to / potentially is not required if this [PR #264](https://github.com/josdejong/workerpool/pull/264) is merged in -> But I was thinking maybe the cleanest solution was to actually ensure that we properly clean up event listeners for the terminated workers, rather than checking if the `WorkerHandler` is `terminated` within the event listener.

I'll leave it up to @josdejong to decide which solution you like better or if you think both PRs should be merged 🤷

### CODE REVIEW

**Comment by hhprogram** [comment]
@josdejong let me know if the unit tests are comprehensive enough, if not, let me know what you think might be missing and I can work to add

**Comment by hhprogram** [comment]
There was an interesting corner case though that I noticed that when I ran the below:

```
      var handler = new WorkerHandler();
      const worker = handler.worker;
      handler.worker = null;
      handler.terminate(true, () => {
        worker.emit('message', 'ready');
        done();
      });
```

I still got the `TypeError: Cannot set property 'ready' of null` , so if you do think this change should be merged - we still might need the change in this #264

**Comment by josdejong** [comment]
Thanks Harrison! This PR looks like it solves the underlying cause, whereas #264 solves the effect. I think this should make #264 redundant, but on the other hand: we're talking here about handling exceptional cases of terminating crashing/unresponsive workers, and weird stuff may happen (like you already encountered seen your comment above). So I think it's a good idea anyway to add the robustness fix of #264 _too_ as an extra measure, and do both. What do you think?

File: src/WorkerHandler.js [hunk 1]
```diff
@@ -400,6 +400,9 @@ WorkerHandler.prototype.terminate = function (force, callback) {
     // all tasks are finished. kill the worker
     var cleanup = function(err) {
       me.terminated = true;
+      if (!(me.worker == null)) {
```
[the remark this answers is not in the document]
Comment (hhprogram) [reply src/WorkerHandler.js:403]: nope no reason - I apparently wanted to do it the hard way. I will change to make it more readable

**Comment by hhprogram** [comment]
> Thanks Harrison! This PR looks like it solves the underlying cause, whereas #264 solves the effect. I think this should make #264 redundant, but on the other hand: we're talking here about handling exceptional cases of terminating crashing/unresponsive workers, and weird stuff may happen (like you already encountered seen your comment above). So I think it's a good idea anyway to add the robustness fix of #264 _too_ as an extra measure, and do both. What do you think?

Ya, that was my original thinking as well too. Do you want me to combine them into one PR or do you just plan to merge them both in?

**Comment by josdejong** [comment]
Sounds good! 

It would be nice if you could add the change from #264 into this PR, that will save me some merge conflicts in the unit test :)

**Comment by hhprogram** [comment]
> Sounds good!
> 
> It would be nice if you could add the change from #264 into this PR, that will save me some merge conflicts in the unit test :)

done!

**Review by josdejong — APPROVED** [review]
Thanks for the latest updates, looks good 👍

**Comment by josdejong** [comment]
Fix published now in version `6.1.3`, thanks again!

**Comment by hhprogram** [comment]
appreciate the responsiveness 🙏
